source/utils/synth.py

```diff
@@ -0,0 +1,31 @@
+# A part of NonVisual Desktop Access (NVDA)
+# Copyright (C) 2023 Beka Gozalishvili
+# This file may be used under the terms of the GNU General Public License, version 2 or later.
+# For more details see: https://www.gnu.org/licenses/gpl-2.0.html
+
+
+import config
+from gui.settingsDialogs import _synthWarningDialog
+import nvwave
+from synthDriverHandler import getSynth, setSynth
+import tones
+
+
+def getOutputDeviceNames():
+	deviceNames = nvwave.getOutputDeviceNames()
+	# #11349: On Windows 10 20H1 and 20H2, Microsoft Sound Mapper returns an empty string.
+	if deviceNames[0] in ("", "Microsoft Sound Mapper"):
+		# Translators: name for default (Microsoft Sound Mapper) audio output device.
+		deviceNames[0] = _("Microsoft Sound Mapper")
+	return deviceNames
+
+
+def setOutputDevice(device: str):
+    config.conf["speech"]["outputDevice"] = device
+    currentSynth = getSynth()
+    if not setSynth(currentSynth.name):

```

Comment:
I'm not sure I agree with the logic here. I understand you haven't written this code, but since we're revisiting it anyway let's try to make it nicer. Currently we set the configured device **before** checking if it can be used i.e. if the current synthesizer can speak through it. Also with the current design there is an import cycle, since you import `gui` here, yet `gui` imports from this new module. I'd move this to `nvwave`, but rewrite the code (in my view basic module for playing sound should not depend on other modules `tones`, synthesizer drivers, which depend on it), using `extensionPoints.Decider` in `nvwave` to which both `tones` and `synthDriverHandler` can subscribe (something like `onMainAudioDeviceChanged`). If both synthesizer and tones can work with the new device, only then it should be set in the config. Also in that case you can check what has been returned from the decider in the settings dialog, and display error in the gui there, and announce the new device, or failure in switching into it in the keyboard scripts. With this approach it would be necessary to modify tones and `setSynth` to accept the specified audio device, as until now they were always using the default one.

Reply:
hi.

Could you please elaborate more on this?

Reply:
I have looked at the code more closely and while I still believe that setting device in the config *before* reinitializing synthesizer is not ideal, changing this will require a huge refactoring. Doing this as part of this PR is probably unjustified, so I'm fine with the current approach.